orca-redis/src/main/java/com/netflix/spinnaker/orca/locks/RedisLockManager.java

```diff
@@ -0,0 +1,244 @@
+      this.lockHolder = lockHolder;
+      this.ttlSeconds = ttlSeconds;
+    }
+
+    List<String> key() {
+      return asList(getLockKey(lockName));
+    }
+
+    List<String> acquireArgs() {
+      return asList(lockValue.getApplication(), lockValue.getType(), lockValue.getId(), lockHolder, Integer.toString(ttlSeconds));
+    }
+
+    List<String> extendArgs() {
+      return asList(lockValue.getApplication(), lockValue.getType(), lockValue.getId(), Integer.toString(ttlSeconds));
+    }
+
+    List<String> releaseArgs() {
+      return asList(lockValue.getApplication(), lockValue.getType(), lockValue.getId(), lockHolder);
+    }
+  }
+
+  private void checkResult(LockOperation op, List<String> result) {
+    final LockValue currentLockValue = buildResultLockValue(result);
+    if (!(op.lockValue.equals(currentLockValue))) {
+      throw new LockFailureException(op.lockName, currentLockValue);
+    }
+  }
+
+  private LockValue buildResultLockValue(List<String> result) {
+    if (result == null || result.size() < 3) {
+      throw new IllegalStateException("Unexpected result from redis: " + result);
+    }
+    if (result.stream().allMatch(Objects::isNull)) {
+      return null;
+    }
+    return new LockValue(result.get(LOCK_VALUE_APPLICATION_IDX), result.get(LOCK_VALUE_TYPE_IDX), result.get(LOCK_VALUE_IDX));
+  }
+
+  private void withLockingConfiguration(LockOperation lockOperation, Consumer<LockOperation> lockManagementOperation) throws LockFailureException {
+    if (!lockingConfigurationProperties.isEnabled()) {
+      return;
+    }
+    try {
+      lockManagementOperation.accept(lockOperation);
+    } catch (Throwable t) {
+      if (t instanceof LockFailureException) {
+        LockFailureException lfe = (LockFailureException) t;
+        Optional<LockValue> currentLockValue = Optional.ofNullable(lfe.getCurrentLockValue());
+        log.info("LockFailureException during {} for lock {} currently held by {} {} {} requested by {} {} {} {}",
+          kv("operationName", lockOperation.operationName),
+          kv("lockName", lockOperation.lockName),
+          kv("currentLockValue.application", currentLockValue.map(LockValue::getApplication).orElse(null)),
+          kv("currentLockValue.type", currentLockValue.map(LockValue::getType).orElse(null)),
+          kv("currentLockValue.id", currentLockValue.map(LockValue::getId).orElse(null)),
+          kv("requestLockValue.application", lockOperation.lockValue.getApplication()),
+          kv("requestLockValue.type", lockOperation.lockValue.getType()),
+          kv("requestLockValue.id", lockOperation.lockValue.getId()),
+          kv("requestLockHolder", Optional.ofNullable(lockOperation.lockHolder).orElse("UNSPECIFIED")),
+          lfe);
+        if (lockingConfigurationProperties.isLearningMode()) {
+          return;
+        }
+        throw lfe;
+      } else {
+        log.info("Exception during {} for lock {} requested by {} {} {} {}",
```

Comment:
`info` the right level for this?

Reply:
I'll switch to debug - if we are in learning mode then this is not a fatal thing, if we are not in learning mode then we'll propagate an exception out of here and fail elsewhere